### ISSUE TITLE
Feature proposal: fail on empty givens

### ISSUE THREAD

**Issue by knittl** [TNG/ArchUnit#532]
It would be awome if ArchUnit could be instructed to fail on "empty givens", i.e. specifying a rule that does not match any code unit. Otherwise it is too easy to have rules giving a false sense of security: they succeed, even though they should not. This is especially important/useful for "negative tests", i.e. "all methods must not …". But it is also useful for existing tests/rules which could break through sloppy refactoring. Consider the following simple test to assert that all methods with a specific annotation have to follow a naming-pattern:

```
@AnalyzeClasses(packages = "com.example")
public final class ArchTests {
    @ArchTest
    static final ArchRule meta = methods()
            .that().areMetaAnnotatedWith("MyAnnotation")
            .should().haveNameEndingWith("Internal");
}
```

This works fine. The test initially fails and the code then fixed to match the rule. Afterwards, the classes containing the methods are moved to another package or the annotation is renamed (I'm aware that there is a `Class<?>` overload for specifying the annotation, this is mainly for illustrative purposes). The test still succeeds, but only because `methods().that().…` returns 0 methods to validate with the rule :zap:

It would be helpful if this could be expressed with ArchUnit somehow: _fail the rule if no code units are selected_.
I'm not sure what the best place for this is; a class-level annotation, a property on the `@ArchTest` annotation, or a method in the DSL (or a combination thereof). Unfortunately, I cannot think of a good name (or place) in the DSL :slightly_frowning_face: The best I can come up with up until now is: `methods().that().areMetaAnnotated("…").mustExist().should()…`

**Comment by hankem** [comment]
FYI: For bettter refactoring safetly, I try to avoid the _stringly typed_ API as much as possible.
In this case, I'd use `@AnalyzeClasses(packagesOf = MyAnnotation.class)` and [`areMetaAnnotatedWith(MyAnnotation.class)`](https://javadoc.io/doc/com.tngtech.archunit/archunit/latest/com/tngtech/archunit/lang/syntax/elements/MembersThat.html#areMetaAnnotatedWith(com.tngtech.archunit.base.DescribedPredicate)) – but your point is still valid,  of course.

**Comment by knittl** [comment]
@hankem yes, I should have mentioned that this was only for the sake of the example. The same problem applies to moved packages (but that's less common, I assume).

In fact, I was stumbling onto this issue when I used an old version of ArchUnit which did not support meta-annotations out of the box (meaning you had to manually add the annotation's package to the scan). Tests were looking good, but were not asserting anything ("∀x∈∅: x > 42" is technically true).

I'm contemplating if this should even be the default behavior: why have a rule that does not apply to anything?

**Comment by hankem** [comment]
One use case that may require rules not matching classes is one where a set of rules for the architecture of a larger system is defined in a central module, but applied to multiple modules, where not all rules actually match in all modules.

**Comment by codecholeric** [comment]
Sounds reasonable, but you can already add this yourself, if I understand you correctly. It's a little hacky, but I think this would do the job:

```
@AnalyzeClasses(packages = "com.example")
public final class ArchTests {
    @ArchTest
    static final ArchRule meta = methods()
            .that().areMetaAnnotatedWith("MyAnnotation")
            .should().haveNameEndingWith("Internal")
            .andShould(exist());

    private <T> ArchCondition<T> exist() {
        return new ArchCondition<T>("exist") {
            @Override
            public void init(Iterable<T> allObjectsToTest) {
                if (Iterables.isEmpty(allObjectsToTest)) {
                    throw new AssertionError("rule does not apply to any object");
                }
            }
    
            @Override
            public void check(T item, ConditionEvents events) {
                // irrelevant for this condition
            }
        };
    }
}
```

Of course we could also add this to the standard API, I think it makes sense i.A. Or could also think about a configuration flag that would fail on any empty rule so you don't have to add it to each rule :thinking: (but I think this must be configurable, because some people might want to slap their whole rule suite on various projects, even if some rules don't apply :thinking:)

**Comment by knittl** [comment]
Good morning @codecholeric oh yes, that looks like it works. I knew I was missing something :slightly_smiling_face: I have pondering a bit more about this over night and maybe I will reformulate this to `atLeast(n)`, maybe I expect a certain number of code units to be checked.

Regarding the configuration flag, I wonder what a good default would be: on or off :thinking: "off" being the current behavior of course

Another possibility I could think of is checking the `ConditionEvents` instance after evaluating a rule. It should contain failed and succeeded checks. If both are 0, then the Condition is either implemented incorrectly™ or nothing was checked. Or is it not required for conditions to add successful checks to the `events` instance?

**Comment by codecholeric** [comment]
About checking the condition events, I actually want to refactor this when I get a chance to not retain successful events :wink: This was originally done so it could easily be negated later, but actually it needs a lot of memory for nothing.

Probably "on" would be a good value, since I consider the harm of this running without any actual check by accident bigger than this complaining even though it justifiedly checked nothing :man_shrugging: Would have to write it in the release notes. But I'm not sure when I'll get around to this, because I also don't think it's completely trivial to patch this in in a backwards compatible way :thinking:

**Comment by thmuch** [comment]
I think Mockito made a similar move – from 3.0 on unused stubs will fail validation automatically. But you still [can allow](http://blog.mockito.org/2018/07/new-mockito-api-lenient.html) unused stubs explicitely using ```lenient()``` at the beginning of the stubbing.

In ArchUnit, this might look like ```ArchRule rule = lenient().classes().that()...```

Maybe we could refactor ```ArchRuleDefinition.Creator``` for this use case (where ```priority()``` is located)? And for being backwards compatible, I'd expect a property like ```rules.lenientByDefault``` (with a default of ```false```).

Would be a great change IMHO, as empty givens can cause lots of frustration late in the game.

**Comment by codecholeric** [comment]
Yeah, that might be a good way to go forward! I agree that unintended empty givens are kind of dangerous and frustrating :+1:

**Comment by codecholeric** [comment]
Only downside of integrating it into the fluent API is that other APIs always need to be adjusted (e.g. `layeredArchitecture`), as that doesn't share the same base rule syntax infrastructure...

**Comment by thmuch** [comment]
Would be nice indeed if there were only a few places that'd need to be changed... I'll have a look at the code if it seems feasible to have a decorator like ```FreezingArchRule.freeze()```, something like ```LenientArchRule.lenient( classes().that()... )```. What do you think?

If we cannot settle down on the API yet, another way to begin with might me the property (i.e. the general on/off switch) with a backward compatible default. Or should we better wait for the refactoring of successful events that you mentioned above?

**Comment by codecholeric** [comment]
The decorator definitely would be a way to add this in a generic way that would not need all APIs to be adjusted :thinking: I'm just not sure if it's feasible in the end, but if, then it might be a good idea. Since in the end you only need it anyway to mix different types, otherwise you can use the generic setting. So only if you want some lenient and some strict rules you would need something like this, otherwise configure the setting globally and be done with it :thinking:

**Comment by codecholeric** [comment]
I think this has been solved by #774 :slightly_smiling_face: If not feel free to reopen!

**Comment by knittl** [comment]
I tested it with one of my initial tests and it is now failing:

> […] failed to check any classes. This means either that no classes have been passed to the rule at all, or that no classes passed to the rule matched the `that()` clause. To allow rules being evaluated without checking any classes you can either use `ArchRule.allowEmptyShould(true)` on a single rule or set the configuration property `archRule.failOnEmptyShould = false` to change the behavior globally.

Old behavior can be restored with `.allowEmptyShould(true)`.

Perfect. Thank you very much! :heart:
